[msbuild] Localization String Test for All Strings

tj-devel709 opened the pull request:
This test will run through all available strings in MSBStrings.Designer.cs in the each of the supported languages.
If any strings should fire in other languages but are still in english, an error will be thrown unless that error code is in an ignore file.

There are two types of ignore files introduced: 
- `LocalizationIgnore/common-Translations.ignore`
    - for error codes that are not yet translated in all languages
- `LocalizationIgnore/<LOCALE>-Translations.ignore`
    - for error codes that are not yet translated in the specific locale

If an error code is in an ignore file and it turns out it is translated, an error will be thrown to remove the error code from the ignore file.

Wiki will be updated for the team to reflect this test: https://github.com/xamarin/maccore/wiki/Localization

*open to feedback on new directory and ignore file names/format

monojenkins commented:
Build failure
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15589//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15589//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15589//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15589//API_20diff_20_28PR_20only_29) (no change)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15589//Generator_20Diff) (no change)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15589//Test_20Report) 🔥

# Test results

<details>
<summary>1 tests failed, 52 tests passed.</summary>

## Failed tests

 * MSBuild tests/iOS (tasks): Failed (Execution failed with exit code 13)
</details>

spouliot commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +84,77 @@ private string TranslateError (string culture, string errorCode)
+
+			foreach (var errorCodeInfo in typeof (MSBStrings).GetProperties ()) {
+				try {
+					errorCode = errorCodeInfo.Name.ToString ();
```
why the _extraneous_ `.ToString ()` call ?
looks like an `PropertyInfo` instance, so `Name` is already a `string` right ?

spouliot commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +84,77 @@ private string TranslateError (string culture, string errorCode)
+					if (errorCode == "ResourceManager" || errorCode == "Culture")
+						continue;
+
+					Assert.IsFalse (string.IsNullOrEmpty (errorCode), "Error code is null or empty");
```
a `PropertyInfo.Name` cannot be `null` (or empty)

spouliot commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +84,77 @@ private string TranslateError (string culture, string errorCode)
+
+					if (commonIgnoreList.Contains (errorCode)) {
+						if (englishError != newCultureError)
+							Assert.IsFalse (true, $"{errorCode} is translated. Remove {errorCode} from ~/LocalizationIgnore/common-Translations.ignore");
```
I think what you're looking for is `Assert.Fail`

rolfbjarne replied:
Or `Assert.AreNotEqual (englishError, newCultureError, $"...")` - the advantage of this being that both `englishError` and `newCultureError` will show up in the logs.

spouliot commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +84,77 @@ private string TranslateError (string culture, string errorCode)
+							Assert.IsFalse (true, $"{errorCode} is translated. Remove {errorCode} from ~/LocalizationIgnore/common-Translations.ignore");
+					} else if (cultureIgnoreList.Contains (errorCode)) {
+						if (englishError != newCultureError)
+							Assert.IsFalse (true, $"{errorCode} is translated. Remove {errorCode} from ~/LocalizationIgnore/{culture}-Translations.ignore");
```
same

spouliot commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +84,77 @@ private string TranslateError (string culture, string errorCode)
+			Assert.AreEqual (string.Empty, errorList, $"The following errors were not translated. Add them to ~/LocalizationIgnore/common-Translations.ignore or ~/LocaliztionIgnore/{culture}-Translations.ignore");
+		}
+
+		private List<string> ReadFile (string path)
```
use `File.ReadAllLines` instead (in the method above)
and just skip lines that starts with `#`

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +84,77 @@ private string TranslateError (string culture, string errorCode)
+				}
+			}
+
+			Assert.AreEqual (string.Empty, errorList, $"The following errors were not translated. Add them to ~/LocalizationIgnore/common-Translations.ignore or ~/LocaliztionIgnore/{culture}-Translations.ignore");
```
This looks like I have to add these files to my home directory, which is not quite right.

mandel-macaque commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +84,77 @@ private string TranslateError (string culture, string errorCode)
+			CultureInfo originalCulture = Thread.CurrentThread.CurrentUICulture;
+			string path = $"{Directory.GetCurrentDirectory ()}/TaskTests/LocalizationIgnore";
+
+			List<string> commonIgnoreList = ReadFile ($"{path}/common-Translations.ignore");
+			List<string> cultureIgnoreList = ReadFile ($"{path}/{culture}-Translations.ignore");
```
I don't like the fact that we are adding empty files with a comment, I would prefer for the test to be written so that the file can be missing and not fail.

mandel-macaque replied:
This two lists can be a Set that is the resulting union of the ignored errors in the common file and in the current translation. What is more, there is no need to read the common ignore list for each of the cultures, we should do this once and re-use the value.

mandel-macaque commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +84,77 @@ private string TranslateError (string culture, string errorCode)
+					if (commonIgnoreList.Contains (errorCode)) {
+						if (englishError != newCultureError)
+							Assert.IsFalse (true, $"{errorCode} is translated. Remove {errorCode} from ~/LocalizationIgnore/common-Translations.ignore");
+					} else if (cultureIgnoreList.Contains (errorCode)) {
```
We are doing the "Contains" for the same error twice. This is not needed. Rather than using two lists, use a single one that is the union of the common error and the culture specific one. That way, we remove the two contains calls and also the twi different assertions. We just need this to be done once.

rolfbjarne replied:
I thought about this too, but the assert message is different, so I think it's worth it to keep the two checks.

tj-devel709 replied:
I thought about this issue too and thought it would be helpful to show where the error to remove is.

mandel-macaque commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +84,77 @@ private string TranslateError (string culture, string errorCode)
+		{
+			List<string> ignoreList = new List<string> ();
+			string temp;
+			try {
+				using (var sr = new StreamReader (path)) {
+					while (!sr.EndOfStream){
+						if (sr.Peek () == -1)
+							break;
+						temp = sr.ReadLine ();
+						if (string.IsNullOrEmpty (temp) || temp.Trim ()[0] == '#') {
+							continue;
+						} else {
+							ignoreList.Add (temp.Trim ());
+						}
+					}
+				}
+			} catch (IOException e) {
+				Assert.IsTrue (false, $"Could not read file at: {path}");
+			}
```
+1 on @spouliot comment. If you wanted to use a Stream, and for memory reasons go line by line, this is an overkill. You can do the following:
```csharp
using (var reader = new StreamReader (path)) {
	string line;
	while ((line = reader.ReadLine ()) != null) {
		if (line.StartsWith ("#"))
				continue;
		else
				ignoreList.Add (line.Trim ());
	}
}
```

rolfbjarne replied:
Or even a one-liner:

```csharp
return File.ReadAllLines (path).Where (line => !line.StartsWith ("#", StringComparison.Ordinal));
```

mandel-macaque replied:
@rolfbjarne are we playing  ⛳  ?

mandel-macaque commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +84,77 @@ private string TranslateError (string culture, string errorCode)
+		}
+
+		private List<string> ReadFile (string path)
+		{
```
What if path is not present, we will have an exception. Do a File.Exist, much nicer over doing work that is no needed.

rolfbjarne replied:
You can return an empty array here if the file does not exist, which simplifies the code above:

```csharp
if (!File.Exists (path))
    return Array.Empty<string> ();
```

mandel-macaque reviewed (changes requested):
I really dis-like adding empty files with not ignored errors. Both in xtro and the bcl tests we support the fact that the file is not present. The two list can be a single union Set and since we are doing that, we should be able to raise a warning of some type stating that an error is ignored in the common file and in the language specific one like we do in the xtro tests.

tj-devel709 commented:
@spouliot @mandel-macaque @rolfbjarne 
I believed I fixed the changes you guys requested! 
I also added two additional lists that track which error codes were found and can report back if there is anything in the ignore files that are not valid error codes.

monojenkins commented:
Build failure
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15606//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15606//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15606//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15606//API_20diff_20_28PR_20only_29) (no change)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15606//Generator_20Diff) (no change)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15606//Test_20Report) 🔥

# Test results

<details>
<summary>1 tests failed, 52 tests passed.</summary>

## Failed tests

 * MSBuild tests/iOS (tasks): Failed (Execution failed with exit code 13)
</details>

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,76 @@ private string TranslateError (string culture, string errorCode)
 			return (string) propertyInfo.GetValue (null, null);
 		}
+
+		private List<string> commonIgnoreList { get; set; } = null;
```
We don't use `private` when it's the default:

```suggestion
		List<string> commonIgnoreList { get; set; } = null;
```

mandel-macaque replied:
Do we even need a property at all? Can be an instance variable.

tj-devel709 replied:
I made this one a property so we could update it the first time and then use the values in for the other tests

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,76 @@ private string TranslateError (string culture, string errorCode)
+
+			if (commonIgnoreList == null && isCommonFileExistant){
+				commonIgnoreList = new List<string> (ReadFile (fullCommonPath));
+				isFirstRun = true;
```
This variable is not needed, it's never read.

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,76 @@ private string TranslateError (string culture, string errorCode)
+			Assert.AreEqual (new List<string> (), commonIgnoreList.Except (commonValidEntries), $"Not all error codes in {shortCommonPath} are valid. Please remove invalid codes");
+		}
+
+		private IEnumerable<string> ReadFile (string path)
```
```suggestion
		IEnumerable<string> ReadFile (string path)
```

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,76 @@ private string TranslateError (string culture, string errorCode)
+			bool isCultureFileExistant = File.Exists (fullCulturePath);
+			CultureInfo originalCulture = Thread.CurrentThread.CurrentUICulture;
+
+			if (commonIgnoreList == null && isCommonFileExistant){
+				commonIgnoreList = new List<string> (ReadFile (fullCommonPath));
+				isFirstRun = true;
+			}
```
If:

1. ReadFile is changed to handle non-existent files
2. ReadFile is changed to return `IList<string>`
3. The `cultureIgnoreList` is changed to be an `List<string>`

This can be simplified to:

```suggestion
			if (commonIgnoreList == null)
				commonIgnoreList = ReadFile (fullCommonPath);
```

tj-devel709 replied:
@rolfbjarne I am running into casting issues when trying to follow the above. Regarding Step 3, cultureIgnoreList is already a List<string>. Was this possibly a typo? Thanks!

tj-devel709 replied:
So I was able to return IList<string> from ReadFile by appending .ToList() to the return statement and changing Array.Empty<string> () to new List<string> (). 
I also casted the ReadFile calls to make them "ReadFile (fullCommonPath) as List<string>"
Is there a better way to do that?

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,76 @@ private string TranslateError (string culture, string errorCode)
+				isFirstRun = true;
+			}
+
+			if (isCultureFileExistant)
+				cultureIgnoreList = new List<string> (ReadFile (fullCulturePath));
```
and:

```suggestion
			cultureIgnoreList = ReadFile (fullCulturePath);
```

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,76 @@ private string TranslateError (string culture, string errorCode)
+					string englishError = TranslateError ("en-US", errorCode);
+					string newCultureError = TranslateError (culture, errorCode);
+
+					if (isCommonFileExistant && commonIgnoreList.Contains (errorCode)) {
```
You don't need the isCommonFileExistant check here.
```suggestion
					if (commonIgnoreList.Contains (errorCode)) {
```

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,76 @@ private string TranslateError (string culture, string errorCode)
+					if (isCommonFileExistant && commonIgnoreList.Contains (errorCode)) {
+						Assert.AreEqual (englishError, newCultureError, $"{errorCode} is translated. Remove {errorCode} from {shortCommonPath}");
+						commonValidEntries.Add (errorCode);
+					} else if (isCultureFileExistant && cultureIgnoreList.Contains (errorCode)) {
```
and:

```suggestion
					} else if (cultureIgnoreList.Contains (errorCode)) {
```

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,76 @@ private string TranslateError (string culture, string errorCode)
+			}
+
+			Assert.AreEqual (string.Empty, errorList, $"The following errors were not translated. Add them to {shortCommonPath} or {shortCulturePath}");
+			Assert.AreEqual (new List<string> (), cultureIgnoreList.Except (cultureValidEntries), $"Not all error codes in {shortCulturePath} are valid. Please remove invalid codes");
```
I think this looks a bit better:

```suggestion
			Assert.IsEmpty (cultureIgnoreList.Except (cultureValidEntries), $"Not all error codes in {shortCulturePath} are valid. Please remove invalid codes");
```

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,76 @@ private string TranslateError (string culture, string errorCode)
+
+			Assert.AreEqual (string.Empty, errorList, $"The following errors were not translated. Add them to {shortCommonPath} or {shortCulturePath}");
+			Assert.AreEqual (new List<string> (), cultureIgnoreList.Except (cultureValidEntries), $"Not all error codes in {shortCulturePath} are valid. Please remove invalid codes");
+			Assert.AreEqual (new List<string> (), commonIgnoreList.Except (commonValidEntries), $"Not all error codes in {shortCommonPath} are valid. Please remove invalid codes");
```
Same

spouliot reviewed (approved):
once @rolfbjarne comments are addressed

mandel-macaque commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,76 @@ private string TranslateError (string culture, string errorCode)
+		[TestCase ("zh-TW")]
+		public void AllErrorTranslation (string culture)
+		{
+			string errorCode = string.Empty;
```
this var is used in the loop at https://github.com/xamarin/xamarin-macios/pull/9971/files#diff-6fa1d6c0dbf0fad08398e2fc9ed0b96550f0000bd031dbe8620035340d2b39d1R131 why do we init it here? Can we just do:
```csharp
var errorCode = errorCodeInfo.Name
```
I don't see it used in other scope but the loop.

mandel-macaque commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,76 @@ private string TranslateError (string culture, string errorCode)
+		public void AllErrorTranslation (string culture)
+		{
+			string errorCode = string.Empty;
+			string errorList = string.Empty;
```
Same.

tj-devel709 replied:
errorCode is used only inside the foreach loop, but errorList is used after the loop to report which errorCodes were not translated

mandel-macaque replied:
WE could use a StringBuilder rather that += "{errorcode} "

mandel-macaque commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,76 @@ private string TranslateError (string culture, string errorCode)
+			string fullCulturePath = $"{Directory.GetCurrentDirectory ()}/TaskTests/LocalizationIgnore/{culture}-Translations.ignore";
+			string shortCommonPath = "xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/Tasktests/LocalizationIgnore/common-Translations.ignore";
+			string shortCulturePath = $"xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/Tasktests/LocalizationIgnore/{culture}-Translations.ignore";
+			bool isCommonFileExistant = File.Exists (fullCommonPath);
+			bool isCultureFileExistant = File.Exists (fullCulturePath);
```
If we follow @rolfbjarne comment, were we return an empty list and the ReadFile supports missing files, we don't need these variables AFAIK.

tj-devel709 commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,76 @@ private string TranslateError (string culture, string errorCode)
+			CultureInfo originalCulture = Thread.CurrentThread.CurrentUICulture;
+
+			if (commonIgnoreList == null && isCommonFileExistant){
+				commonIgnoreList = new List<string> (ReadFile (fullCommonPath));
```
@mandel-macaque i tried doing the common ignore in a [SetUp]. Is this how you were envisioning?

tj-devel709 replied:
Also can you explain why it is thread unsafe? Is it because I am trying to read two different files back to back?

monojenkins commented:
Build failure
❌ [Build failed or was aborted](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15626/)

✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15626//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15626//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15626//API_20diff_20_28from_20stable_29)

monojenkins commented:
Build failure
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15627//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15627//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15627//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15627//API_20diff_20_28PR_20only_29) (no change)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15627//Generator_20Diff) (no change)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15627//Test_20Report) 🔥

# Test results

<details>
<summary>1 tests failed, 52 tests passed.</summary>

## Failed tests

 * MSBuild tests/iOS (tasks): Failed (Execution failed with exit code 13)
</details>

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,74 @@ private string TranslateError (string culture, string errorCode)
 			return (string) propertyInfo.GetValue (null, null);
 		}
+
+		List<string> commonIgnoreList = null;
```
You can declare this as IList:

```suggestion
		IList<string> commonIgnoreList = null;
```

that will remove the need for casting to `List<string>` and the `.ToList ()` call later.

tj-devel709 replied:
Switching the commonIgnoreList and cultureIgnoreList to IList<string> worked, but if I take off the .ToList(), I get this compiler error 
"TaskTests/LocalizationStringTest.cs(154,11): error CS0266: Cannot implicitly convert type 'System.Collections.Generic.IEnumerable<string>' to 'System.Collections.Generic.IList<string>'. An explicit conversion exists (are you missing a cast?) [/Users/tjlambert/Documents/AllStrings/xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/Xamarin.MacDev.Tasks.Tests.csproj]"

whitneyschmidt replied:
Hm...if you look up the C# docs what is the relationship between IEnum and IList? It looks like you probably need some kind of a cast: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/types/casting-and-type-conversions

tj-devel709 replied:
I had also tried appending "... as IList<string>" instead of ".ToList()". It seemed that it was returning null since I was getting this NullReferenceException when trying to run the .Contains() method on commonIgnoreList in line 132. Perhaps I am doing the casting incorrectly though 

13) Error : Xamarin.iOS.Tasks.LocalizationStringTest.AllErrorTranslation("zh-TW")
System.NullReferenceException : Object reference not set to an instance of an object
  at Xamarin.iOS.Tasks.LocalizationStringTest.AllErrorTranslation (System.String culture) [0x000c8] in /Users/tjlambert/Documents/AllStrings/xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:132 
  at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
  at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-06/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Reflection/RuntimeMethodInfo.cs:395

tj-devel709 replied:
Also if I try (IList<string>) as a cast, I get this:
System.InvalidCastException : Specified cast is not valid.
  at Xamarin.iOS.Tasks.LocalizationStringTest.ReadFile (System.String path) [0x00016] in /Users/tjlambert/Documents/AllStrings/xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:154 
  at Xamarin.iOS.Tasks.LocalizationStringTest.SetUp () [0x00001] in /Users/tjlambert/Documents/AllStrings/xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:94 
  at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
  at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in /Users/builder/jenkins/workspace/build-package-osx-mono/2019-06/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Reflection/RuntimeMethodInfo.cs:395

mandel-macaque replied:
You do not need a IList, and IList and IEnumerable are diff interfaces, but an IList is a IEnumerable as per inheriting from ICollection. You should not cast since it wont work (or there is no guarantee it will). If you were to remove the IList, you will have to do a number of changes:

1. ReadFile -> Should return an IEnumerable, we remove the call toList
2. You will need to use the extension method for IEnumerable to do Contains, since IEnumerables only have GetEnumerator, that method is here: https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.contains?view=netcore-3.1 which is present in Linq, means you will need the name space to be added.

@tj-devel709 @whitneyschmidt and IEnumerable is an interface that will only return an IEnumerator<out T>. That allows you to create "collections" that you cannot insert or remove from but iterate. This collections can be used in a foreach loop.

You can see here that the IL generated is a little diff https://sharplab.io/#gist:26a263aebdb501b064d4f022be8db657

IEnumerables are very useful when working with collections that you generated on the fly, with yield return. @tj-devel709  you probably have seen a lot of yield returns in binding tools for swift.

```csharp
public class Galaxies
    {

        public System.Collections.Generic.IEnumerable<Galaxy> NextGalaxy
        {
            get
            {
                yield return new Galaxy { Name = "Tadpole", MegaLightYears = 400 };
                yield return new Galaxy { Name = "Pinwheel", MegaLightYears = 25 };
                yield return new Galaxy { Name = "Milky Way", MegaLightYears = 0 };
                yield return new Galaxy { Name = "Andromeda", MegaLightYears = 3 };
            }
        }
    }
```

You can also find them in several of our tests that create test data: https://github.com/xamarin/xamarin-macios/blob/main/tests/xharness/Microsoft.DotNet.XHarness.iOS.Shared.Tests/Hardware/DeviceTest.cs#L49

For those that have used python, it is very similar to this:

```python
def csv_reader(file_name):
    for row in open(file_name, "r"):
        yield row
```
When a file is too large, you will run out of memory (in the above example). But with a yield you do not store the entire collection in memory, yet you can iterate through it.

Most of the methods from Linq do return IEnumerables because they are used to work with large collections and instead of duplicating the memory usage they use an IEnumerator with a MoveNext: https://github.com/microsoft/referencesource/blob/master/System.Core/System/Linq/Enumerable.cs#L184

yield return will do the magic for you to have a IEnumerator without the need of the class implementation.

mandel-macaque replied:
I forgot to mention, do not remove the ToList, the contains will have to go through the entire list with no gain.

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,74 @@ private string TranslateError (string culture, string errorCode)
+		IList<string> ReadFile (string path)
+		{
+			if (!File.Exists (path))
+				return new List<string> ();
```
You can return an empty array if you do the other change to the field type at the top:

```suggestion
				return Array.Empty<string> ();
```

this is more efficient (it won't create a new instance every time this method is called).

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationIgnore/common-Translations.ignore:
```diff
@@ -0,0 +1,3 @@
+# Insert Error codes that are waiting on translations below (one per line)
+W0111
+E7069
```
According to the test results you need to add `E7070` here as well.

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,74 @@ private string TranslateError (string culture, string errorCode)
+			List<string> commonValidEntries = new List<string> ();
+			List<string> cultureValidEntries = new List<string> ();
+
+			string fullCommonPath = $"{Directory.GetCurrentDirectory ()}/TaskTests/LocalizationIgnore/common-Translations.ignore";
```
This variable isn't used here anymore, so it should be moved to the SetUp method (and used there instead of duplicating the path creation logic).

tj-devel709 replied:
Should I take it out completely since I will only be using it that one time or is it worth keeping to make the code a little cleaner?

rolfbjarne replied:
Either way works!

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,74 @@ private string TranslateError (string culture, string errorCode)
+
+			string fullCommonPath = $"{Directory.GetCurrentDirectory ()}/TaskTests/LocalizationIgnore/common-Translations.ignore";
+			string fullCulturePath = $"{Directory.GetCurrentDirectory ()}/TaskTests/LocalizationIgnore/{culture}-Translations.ignore";
+			string shortCommonPath = "xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/Tasktests/LocalizationIgnore/common-Translations.ignore";
+			string shortCulturePath = $"xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/Tasktests/LocalizationIgnore/{culture}-Translations.ignore";
```
Incorrect case in the path:

```suggestion
			string shortCommonPath = "xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationIgnore/common-Translations.ignore";
			string shortCulturePath = $"xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationIgnore/{culture}-Translations.ignore";
```

rolfbjarne reviewed (changes requested):
This is getting closer!

monojenkins commented:
Build success
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15637//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15637//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15637//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15637//API_20diff_20_28PR_20only_29) (no change)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15637//Generator_20Diff) (no change)
✅ [Test run succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15637//Test_20Report)

rolfbjarne commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,73 @@ private string TranslateError (string culture, string errorCode)
+		{
+			if (!File.Exists (path))
+				return Array.Empty<string> ();
+			return File.ReadAllLines (path).Where (line => !line.StartsWith ("#", StringComparison.Ordinal) && line != string.Empty).ToList();
```
Just leave the `ToList ()` call, that'll work just fine.

Except style:

```suggestion
			return File.ReadAllLines (path).Where (line => !line.StartsWith ("#", StringComparison.Ordinal) && line != string.Empty).ToList ();
```

mandel-macaque commented on tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationStringTest.cs:
```diff
@@ -83,5 +85,73 @@ private string TranslateError (string culture, string errorCode)
+			List<string> cultureValidEntries = new List<string> ();
+
+			string fullCulturePath = $"{Directory.GetCurrentDirectory ()}/TaskTests/LocalizationIgnore/{culture}-Translations.ignore";
+			string shortCommonPath = "xamarin-macios/tests/msbuild/Xamarin.MacDev.Tasks.Tests/TaskTests/LocalizationIgnore/common-Translations.ignore";
```
Could be a static instance var.

monojenkins commented:
Build success
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15648//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15648//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15648//API_20diff_20_28from_20stable_29)
✅ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15648//API_20diff_20_28PR_20only_29) (no change)
✅ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15648//Generator_20Diff) (no change)
✅ [Test run succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15648//Test_20Report)
